src/foam/dao/index/Index.js

```diff
@@ -15,52 +15,114 @@
 
-/*
- * TODO:
- *  update(oldValue, newValue)
- *  reuse plans
- *  add ability for indices to pre-populate data
- */
+  Each Index subclass also defines an IndexNode class. Index defines
+  the structure of the index, including estimate() to gauge its
+  probable performance for a query, while IndexNode implements the
+  data nodes that hold the indexed items and plan and execute
+  queries. For any particular operational Index, there may be
+  many IndexNode instances:
+
+<pre>
+                 1---------> TreeIndex(id)
+  MDAO: AltIndex 2---------> TreeIndex(propA) ---> TreeIndex(id) -------------> ValueIndex
+        | 1x AltIndexNode    | 1x TreeIndexNode    | 14x TreeIndexNodes         | (DAO size)x ValueIndexNodes
+           (2 alt subindexes)     (14 nodes)             (each has 0-5 nodes)
+</pre>
+  The base AltIndex has two complete subindexes (each holds the entire DAO).
+  The The TreeIndex on property A has created one TreeIndexNode, holding one tree of 14 nodes.
+  Each tree node contains a tail instance of the next level down, thus
+  the TreeIndex on id has created 14 TreeIndexNodes. Each of those contains some number
+  of tree nodes, each holding one tail instance of the ValueIndex at the end of the chain.
 
-/** The Index interface for an ordering, fast lookup, single value,
-  index multiplexer, or any other MDAO select() assistance class. */
+*/
 foam.CLASS({
   package: 'foam.dao.index',
   name: 'Index',
 
+  properties: [
+    {
+      class: 'Class',
+      name: 'tailClass',
+      // TODO: a factory would be better, but Class getter overrides it
+    }
+  ],
   methods: [
-    /** JS-prototype based 'Flyweight' constructor. Creates plain
-      javascript objects that are __proto__'d to a modeled instance. */
-    function create(args) {
-      var c = Object.create(this);
-      args && c.copyFrom(args);
-      c.init && c.init();
-      return c;
+
+    function estimate(size, sink, skip, limit, order, predicate) {
+      /** Estimates the performance of this index given the number of items
+        it will hold and the planned parameters. */
+      return size * size; // n^2 is a good worst-case estimate by default
+    },
+
+
+    function toPrettyString(indent) {
+      /** Output a minimal, human readable, indented (2 spaces per level)
+        description of the index structure */
     },
 
+    function createNode(args) {
+      if ( ! this.tailClass ) this.tailClass = this.cls_.id + 'Node';
+      args = args || {};
+      args.creator = this;
+      return this.tailClass.create(args, this);
+    }
+  ]
+});
+
+/**
+  The index tail interface represents a piece of the index that actually
+  holds data. A tree will create a tail for each node, so one Index will
+  create many tail instances, each operating on part of the data in the DAO.
+
+  For creation speed, do not require or import anything in a tail class.
+  Use the 'creator' property to access requires and imports on the
```

Comment:
The creator is the Index instance that both defines the type and parameters for that part of the index, and contains any shared state or context access the lightweight nodes need.

So for nodes, the creator reference lets them access requires without themselves triggering a context build on themselves, and access estimate() or other methods that don't depend on the actual data in the index.

MyIndex.create() // allowed to do heavy setup operations
MyIndex.createNode() --> MyIndexNode.create() // not allowed to be slow
myNodeInstance.creator.ThingRequired // access to things that were set up by MyIndex 

(this avoids the Progenitor implementation that previously modelled and enforced the lightweightness of nodes)